iso19139 / Add better support for gmd:language containing gmd:LanguageCode and gco:CharacterString

ianwallen opened the pull request:
Change condition so that gmd:LanguageCode/@codeListValue has higher priority than gco:CharacterString Since codeListValue are controlled list they are more reliable. gco:CharacterString may contain something like "eng; US"

Basically adding support for something like the following.

```xml
  <gmd:language>
    <gco:CharacterString>eng; USA</gco:CharacterString>
    <gmd:LanguageCode codeList="http://www.loc.gov/standards/iso639-2/" codeListValue="eng"/>
  </gmd:language>
```

ianwallen commented on schemas/iso19139/src/main/plugin/iso19139/index-fields/language-default.xsl:
```diff
@@ -776,7 +776,7 @@
                store="true" index="true"/>
         <Field name="cl_{concat(local-name(), '_text')}"
-               string="{util:getCodelistTranslation(name(*), string(*/@codeListValue), string($isoLangId))}"
+               string="{util:getCodelistTranslation(name(.), string(*/@codeListValue), string($isoLangId))}"
```
I was getting errors of multiple elements being passed to name() when it was using name(*).  I applied this fix however I don't believe it is the correct fix.  I attempted another fix where I changed */@codeListValue to @codeListValue in the for-each and other references which seems to make more sense to me however however it was generating lots of errors related to translating the values.  

Maybe someone understands why I was getting multiple elements being passed to name().

To test it I mostly just used the multilingual template supplied in the profile and replace the gmd:language

ianwallen commented on schemas/iso19139/src/main/plugin/iso19139/layout/layout.xsl:
```diff
@@ -534,8 +534,8 @@
       <xsl:with-param name="name"
-                      select="if ($isEditing) then concat(*/gn:element/@ref, '_codeListValue') else ''"/>
-      <xsl:with-param name="editInfo" select="*/gn:element"/>
+                      select="if ($isEditing) then concat(./gn:element/@ref, '_codeListValue') else ''"/>
```
Same issue - multiple elements to concat()

fxprunayre commented:
@ianwallen is it encoding used a lot ?

Could this be also used ?
```xml
    <gmd:LanguageCode codeList="http://www.loc.gov/standards/iso639-2/" codeListValue="eng">
   eng; USA   
   </gmd:LanguageCode>
```

ianwallen commented:
@fxprunayre You are correct - I thought I had seen this encoding but I guess it is not that common.

We are currently getting data like the following

```
<gmd:language>
    <gco:CharacterString>eng; CAN</gco:CharacterString>
</gmd:language>
```

and it is causing some issues when determining the language so I thought it would be more accurate if this was a code. So I added the gmd:LanguageCode to resolve this without thinking that the "eng; CAN" can actually be moved into the gmd:LanguageCode as well!

The code I added does add more support for determining the language.  if we receive only a CharacterString with a country code as above (determined by the existance of a ";") it will extract the language code from the string.
And in the event that both (CharacterString and codelist) are added it will simply give higher priority to the code list.

**Regarding your suggestion to use gmd:LanguageCode**
I just did a build of the current master branche and there seems to be a bug with using the suggested gmd:LanguageCode...

- If using 

```
 <gmd:LanguageCode codeList="http://www.loc.gov/standards/iso639-2/" codeListValue="eng">
   eng; CAN   
   </gmd:LanguageCode>
```

Then when the data is saved using the editor it is changed to 
```
<gmd:LanguageCode codeList="http://www.loc.gov/standards/iso639-2/" codeListValue="eng"/>
```
So the string portion is lost even though I did not make any changes other than save the data.

- Related to this, I'm not sure how the editor would handle the change.  If language is changed from eng to fra then it will save 
```
<gmd:LanguageCode codeList="http://www.loc.gov/standards/iso639-2/" codeListValue="fra"/>
```

If it had **eng; CAN** in the string how would the editor know to change this to **fra: CAN**

I guess it could add a rule in the update-fixed-info.xsl similar to the following logic.
Does string contain a ";"? if so assume it is a "lang ; country" code format
If "lang ; country" format then add language code then look for country code in pt_locale and apply the country code.
I don't believe the system currently support having multiple locales for eng; USA and eng: UK so there should not be any issue with getting the country code this way.

Another option is to update the editor to allow selection for the code and and a input box for the text but they can get out of sync if the user does not update both.

Looks like this pull request has the same issue where the code and text can get out of sync.

ianwallen commented:
Will need to do some more changes to get tests working. 
I will also look into fixing the issue with the gmd:LanguageCode losing the values.

ianwallen commented:
I have fixed the code so that the unit test work.

As some of the changes were reverted the following will no longer work due to existing bugs in the code (not caused by this pull request).
  <gmd:language>
    <gco:CharacterString>eng; USA</gco:CharacterString>
    <gmd:LanguageCode codeList="http://www.loc.gov/standards/iso639-2/" codeListValue="eng"/>
  </gmd:language>

If those bugs are fixed then the above should work with this new code.

It does add better support for language strings that contain country codes.
  <gmd:language>
    <gco:CharacterString>eng; USA</gco:CharacterString>
  </gmd:language>

Regarding the gmd:LanguageCode losing the values I created a new pull request #4215 to handle this.

Let me know if there are other issues with this request.

Thank you

josegar74 commented on schemas/iso19139/src/main/plugin/iso19139/index-fields/default.xsl:
```diff
@@ -503,10 +503,14 @@
 
       <!-- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -->
 
-      <xsl:for-each
```
In Data Identification the cardinality is [1..*], with the change will be indexed only the first language. Probably it's the most usual case, but would be better to manage all languages defined.

josegar74 commented on schemas/iso19139/src/main/plugin/iso19139/index-fields/default.xsl:
```diff
@@ -896,9 +900,18 @@
 
     <!-- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -->
 
-    <xsl:for-each select="gmd:language/gco:CharacterString
```
For metadata language cardinality is [0..1], probably in most cases will be defined, but without the loop can be indexing empty data I guess, in case no language defined.

josegar74 commented on schemas/iso19139/src/main/plugin/iso19139/layout/utility-tpl-multilingual.xsl:
```diff
@@ -38,8 +38,17 @@
       <xsl:otherwise>
-        <xsl:value-of select="$metadata/gmd:language/gco:CharacterString|
-       $metadata/gmd:language/gmd:LanguageCode/@codeListValue"/>
+          <xsl:choose>
```
Can't be changed to use `langId_from_gmdlanguage19139`?

ianwallen replied:
It probably can however I would need to add an xsl:include  and I was not sure if that was going to cause any conflicts - I can try it and see if it causes any issues.

josegar74 commented on schemas/iso19139/src/main/plugin/iso19139/present/metadata-utils.xsl:
```diff
@@ -87,6 +87,46 @@
 
+  <!-- iso3code from the supplied gmdlanguage
+       It will prefer LanguageCode if it exists over CharacterString -->
+  <xsl:template name="getLangIdFromGmdlanguage">
```
Can't be changed to use `langId_from_gmdlanguage19139`?

ianwallen replied:
I did noticed the duplication and thought of making it just call one however I was not sure of the dependencies - and I did not want to cause any conflict issues.  I believe we should be able to remove these templates and just use the function version as long as it does not cause any conflicts.  I can try it to see if it causes any noticeable issues.

ianwallen replied:
I have updated the code to remove duplication

ianwallen commented:
@josegar74 
You are correct about the index - I have reverted the changes.  It was only causing issues when both the <gco:CharacterString> and <gmd:LanguageCode> where used at the same time but that is no longer the case for this pull request

ianwallen commented:
If there are still some issues with this pull request could I please get some feedback. 
Thank you.

josegar74 commented:
@ianwallen I'll check today and will merge it.

ianwallen commented:
@josegar74 
Do you think it would be possible to have this land in the 3.10.2 release?

josegar74 commented on web/src/main/webapp/WEB-INF/config-csw.xml:
```diff
@@ -108,7 +108,7 @@
         <parameter name="Language" field="language" type="SupportedISOQueryables">
           <xpath schema="iso19139"
-                 path="gmd:language/gco:CharacterString|gmd:language/gmd:LanguageCode/@codeListValue"/>
+                 path="if (normalize-space(gmd:language/gmd:LanguageCode/@codeListValue) != '') then gmd:language/gmd:LanguageCode/@codeListValue else gmd:language/gco:CharacterString"/>
```
@ianwallen I tested the pull request and this change doesn't work, should be the previous code or if case want to check first the LanguageCode switch the order, but I guess that's not relevant as `gmd:language` will have only `gco:CharacterString` or `gmd:LanguageCode`, but not both.

Please update the code and will merge it, thanks.

ianwallen commented:
@josegar74 
I have removed all changes that were related to codelist/characterstring order.
I have squashed all the commits and created a new pull request (#4572).  (I was not sure if I could push squashed commits to this pull request)
So if you can approve pull request #4572 that would be great.  We can then cancel this pull request.

Thank your

josegar74 commented:
Done, will cherry-pick the changes to 3.10.x also so are available for 3.10.2.

ianwallen commented:
pull request [#4572 ](https://github.com/geonetwork/core-geonetwork/pull/4572) has been merged so closing this request.
